core/trino-main/src/main/java/io/trino/operator/RefreshMaterializedViewOperator.java

```diff
@@ -0,0 +1,149 @@
+        private final Metadata metadata;
+        private final QualifiedObjectName viewName;
+        private boolean closed;
+
+        public RefreshMaterializedViewOperatorFactory(int operatorId, PlanNodeId planNodeId, Metadata metadata, QualifiedObjectName viewName)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.viewName = requireNonNull(viewName, "viewName is null");
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext operatorContext = driverContext.addOperatorContext(operatorId, planNodeId, ValuesOperator.class.getSimpleName());
+            return new RefreshMaterializedViewOperator(operatorContext, metadata, viewName);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new RefreshMaterializedViewOperatorFactory(operatorId, planNodeId, metadata, viewName);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final Metadata metadata;
+    private final QualifiedObjectName viewName;
+    @Nullable
+    private ListenableFuture<?> refreshFuture;
+    private boolean closed;
+
+    public RefreshMaterializedViewOperator(OperatorContext operatorContext, Metadata metadata, QualifiedObjectName viewName)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.viewName = requireNonNull(viewName, "viewName is null");
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        if (refreshFuture == null) {
+            return NOT_BLOCKED;
+        }
+
+        return refreshFuture;
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return false;
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    @Override
+    public void finish()
+    {
+    }
+
+    @Override
+    public void close()
+    {
+        if (refreshFuture != null) {
+            refreshFuture.cancel(false);
+        }
+
+        closed = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        if (closed) {
+            return true;
+        }
+
+        if (refreshFuture == null) {
+            refreshFuture = metadata.refreshMaterializedView(operatorContext.getSession(), viewName);
```

Comment:
I don't think it's as good idea to do this as a side-effect of the `isFinished()` call, which is supposed to be just a state check.

Do it in the `getOutput()` method, instead.

Reply:
This is the last operator in pipeline, therefore `getOutput` is not called. Unfortunately, we rely on `isFinished` called repetitively for last operator to progress operator.

See `io.trino.operator.Driver#processInternal` and comment there:
```
            // If there is only one operator, finish it
            // Some operators (LookupJoinOperator and HashBuildOperator) are broken and requires finish to be called continuously
            // TODO remove the second part of the if statement, when these operators are fixed
            // Note: finish should not be called on the natural source of the pipeline as this could cause the task to finish early
```

I believe the fix could be to switch completely to work processors and call `io.trino.operator.WorkProcessor#process`

Reply:
That's unfortunate. Please add a comment, then, as that's not how it should work.